### PR TITLE
Added Certificate-Sender

### LINKED ISSUE #1085: Automatically Send Custom Named Certificates via Mail
`GSSOC21`

# Aim

- Send custom-named certificates via Mail

# Details
- Adding names of participants to the certificate programmatically from CSV file
- Using SMTP to send mails
- Optimized process for creating & sending certificate

### PR DESCRIPTION
**Pull request by dhanrajdc7**
# Description

Send Bulk Certificates Automatically via Mail.

## Type of change

_Please delete options that are not relevant._

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Documentation Update

## Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have created a helpful and easy to understand `README.md`
- [x] My documentation follows [`Template for README.md`](https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/master/Template%20for%20README.md)
- [x] I have added the project meta data in the PR template.
- [x] I have created the ``requirements.txt`` file if needed.

## Project Metadata

`` If there is no-file/nothing to fill the below fields with, then type: none ``

`` Example: `` If no requirements.txt needed/present, then type none in ``Requirments``.

Category:
- [ ] Calculators
- [ ] AI/ML
- [ ] Scrappers
- [ ] Social_Media
- [ ] PDF
- [ ] Image_Processing
- [ ] Video_Processing
- [ ] Games
- [ ] Networking
- [ ] OS_Utilities
- [x] Automation
- [ ] Cryptography
- [ ] Computer_Vision
- [ ] Fun
- [ ] Others

Title: Certificate Sender

Folder: Certificate-Sender

Requirements: requirements.txt

Script: main.py

Arguments: none

Contributor: dhanrajdc7

Description: Send Bulk Certificates Automatically via Mail.

### CODE REVIEW

File: Certificate-Sender/main.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+from PIL import Image
+from PIL import ImageFont
+from PIL import ImageDraw
+
+
+# Mail Setup
+emailID = "abc1234@gmail.com"  # ADD MAIL ID
+pwd = "xyz@1234"  # ADD PASSWORD
+subject = "🥳 Thank You For Attending! Here is your Certificate 🎉"  # EMAIL SUBJECT
+body = """Hi!\n\nThanks for your active participation in our webinar on "All About Git, GitHub & Open Source"
```
Comment (kaustubhgupta) [inline Certificate-Sender/main.py:18]: Get this data from user via input functions

File: Certificate-Sender/main.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+emailID = "abc1234@gmail.com"  # ADD MAIL ID
+pwd = "xyz@1234"  # ADD PASSWORD
+subject = "🥳 Thank You For Attending! Here is your Certificate 🎉"  # EMAIL SUBJECT
+body = """Hi!\n\nThanks for your active participation in our webinar on "All About Git, GitHub & Open Source" 
+Organized by Dhanraj. So now your wait is over and here is your certificate of participation in the webinar.\n\nAlso, 
+don't forget to tag on LinkedIn with your certificate. we are happy to see you there!\n\nDhanraj: 
+https://www.linkedin.com/in/dhanrajdc7/ """
+
+# Files
+excelFile = "Data.xlsx"  # DATA(NAME & EMAIL)
```
Comment (kaustubhgupta) [inline Certificate-Sender/main.py:24]: get the input for data file path from user

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]
- PR template wrongly updated. Check approved PRs for reference.
- Delete the Data1.xlsx
- The readme file doesn't contain any good information about the script. Check other in the repo or approved PRs. 
- Other changes:

**Comment by kaustubhgupta** [comment]
@dhanrajdc7 update the PR before 2 pm so that we can review the code and merge it timely

**Comment by dhanrajdc7** [comment]
@kaustubhgupta Pls review.

File: Certificate-Sender/.env.sample [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Mail Setup
+EMAIL_ID = "abc1234@gmail.com"  # ADD MAIL ID
+EMAIL_PASSWORD = "xyz@1234"  # ADD PASSWORD
+EMAIL_SUBJECT = "🥳 Thank You For Attending! Here is your Certificate 🎉"  # EMAIL SUBJECT
+CERTIFICATE_EMAIL_BODY = "Hi!
+
+Thanks for your active participation in our webinar on \"All About Git, GitHub & Open Source\" Organized by Dhanraj.
+So now your wait is over and here is your certificate of participation in the webinar.
+Also, don't forget to tag on LinkedIn with your certificate. we are happy to see you there!
```
Comment (kaustubhgupta) [inline Certificate-Sender/.env.sample:9]: I have clearly mentioned in this suggestion that takes these values as input from user via input function: https://github.com/avinashkranjan/Amazing-Python-Scripts/pull/1131#discussion_r641585677

File: Certificate-Sender/.gitignore [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Auto-generated
+.idea/
+.vscode/
+
+# Configuration & Environment variables
+.env
+
+# Virtual Environment
+*env/
```
Comment (kaustubhgupta) [inline Certificate-Sender/.gitignore:9]: not needed, you can remove this

File: Certificate-Sender/README.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+Send Bulk Certificates Automatically via mail.
+
+- send bulk certificaates
+- add names & other custom details on Certificate
+- easy to use
+
+## Setup instructions
+
+- Add all details in `.env` file
```
Comment (kaustubhgupta) [inline Certificate-Sender/README.md:11]: not valid

File: Certificate-Sender/main.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    server.sendmail(email_id, receiver, msg.as_string())
+
+
+# Run Script
+if __name__ == '__main__':
+    load_dotenv()
+    error_list = []
+    error_count = 0
+
+    excel_file = os.getenv('CERTIFICATE_HOLDERS_EXCEL_FILEPATH')
```
Comment (kaustubhgupta) [inline Certificate-Sender/main.py:128]: what is this statement? Get the file path from user via input function

File: Certificate-Sender/main.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+):
+    if not template_file:
+        template_file = os.getenv('CERTIFICATE_TEMPLATE_FILEPATH')
+
+    if not font_file:
+        font_file = os.getenv('CERTIFICATE_NAME_FONT_FILEPATH')
+
+    if not starting_position:
+        starting_position = tuple(map(int, os.getenv(
+            'CERTIFICATE_NAME_STARTING_POSITION').split(',')))
```
Comment (kaustubhgupta) [inline Certificate-Sender/main.py:47]: these inputs need to be obtained from user via input function

File: Certificate-Sender/main.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        email_id = os.getenv('EMAIL_ID')
+
+    if not pwd:
+        pwd = os.getenv('EMAIL_PASSWORD')
+
+    if not subject:
+        subject = os.getenv('EMAIL_SUBJECT')
+
+    if not body:
+        body = os.getenv('CERTIFICATE_EMAIL_BODY')
```
Comment (kaustubhgupta) [inline Certificate-Sender/main.py:92]: these too

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]
- Not deleted data1.xlsx
- Other changes:

**Comment by dhanrajdc7** [comment]
@kaustubhgupta 

I have pushed my code by testing & made it very simple for users. If I implement input() & let's say if the program crashes due to Gmail SMTP error (like Daily Limit), then the user will have to enter all details again to run it again. This bad user experience doesn't make any sense!!!! 

I have added Excel File to let know users about the Data format.

**Comment by kaustubhgupta** [comment]
> @kaustubhgupta
> 
> I have pushed my code by testing & made it very simple for users. If I implement input() & let's say if the program crashes due to Gmail SMTP error (like Daily Limit), then the user will have to enter all details again to run it again. This bad user experience doesn't make any sense!!!!
> 
> I have added Excel File to let know users about the Data format.

We are making a master script where your program will be part of a pool of scripts. As soon as the master script is run and the user selects your script, how will the user provide the data? You can't instruct the user to close the master script, go back to the file directory, edit the .env file, and then run the master script! That's why you were told to make changes to the script to make it dynamic. The data format for excel can be explained in the readme file. 

Also, as the GSSOC period is over, only approved PRs will be merged and no new reviews/approvals would be made. You were told to make the changes two days back (31st May).

**Pull request closed without merging**